### PR TITLE
[READY]Jailbreak & detain objectives

### PR DESCRIPTION
**Pull request by itseasytosee**
## About The Pull Request

Adds an objective for making sure a target escapes alive and out of custody.
Currently its admin only.

## Why It's Good For The Game
The idea of breaking a prisoner out of jail as an objective has been tossed around this is the first step.

## Changelog
:cl:
add: Admins can now play around with jailbreak & detain objectives. Which means help a target escape out of custody or insure a target escapes in custody respectively.
/:cl:

### CODE REVIEW

**Comment by HomocidlePancakes** [comment]
Should alert the prisoner who is being rescued.

**Comment by itseasytosee** [comment]
I had the same idea. But its a lot easier to say a thing than to make it happen.

**Comment by itseasytosee** [comment]
The only way I could even fathom making it work would be having it be a team antag thing.

File: code/game/gamemodes/objective.dm [hunk 1]
```diff
@@ -306,6 +306,29 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
 	name = "protect nonhuman"
 	human_check = FALSE
 
+/datum/objective/jailbreak
+	name = "jailbreak"
+	var/target_role_type=FALSE
+	martyr_compatible = 1 //why not?
```
Comment (Cyberboss) [inline code/game/gamemodes/objective.dm]: Move this line up one and set it to `TRUE`

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/gamemodes/objective.dm (hunk 1 above)
Comment (itseasytosee) [reply code/game/gamemodes/objective.dm]: which bit?

Comment (ArcaneMusic) [reply code/game/gamemodes/objective.dm]: He means the Martyr_Comparible var, and it reads better to have new vars for any subtype below the parent vars.

Comment (itseasytosee) [reply code/game/gamemodes/objective.dm]: Oh thanks

**Comment by Cyberboss** [comment]
`true` is not a creative constant

File: code/game/gamemodes/objective.dm (hunk 1 above)
Comment (itseasytosee) [reply code/game/gamemodes/objective.dm]: fixed

**Comment by itseasytosee** [comment]
I don't see why it had to be true instead of 1 it worked fine. but hey I'm no master.

**Comment by MrDoomBringer** [comment]
>I had the same idea. But its a lot easier to say a thing than to make it happen.

What? Unless I'm missing something, I'm pretty sure the feature @HomocidlePancakes mentioned would be very easy to implement (like 3 lines of code). Maybe a bit more if you wanted it to work for any arbitrary objective.

**Comment by itseasytosee** [comment]
how would you go about doing that?

**Comment by davethwave** [comment]
I would say to have both the prisoner and person with objective know of each other or something like that at round start

**Comment by MrDoomBringer** [comment]
Just do it in the objective's find_target function

File: code/game/gamemodes/objective.dm [hunk 2]
```diff
@@ -306,6 +306,29 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
...
 
+/datum/objective/jailbreak
+	name = "jailbreak"
+	martyr_compatible = TRUE //why not?
+	var/target_role_type=FALSE
+
+/datum/objective/jailbreak/find_target_by_role(role, role_type=FALSE,invert=FALSE)
+	if(!invert)
+		target_role_type = role_type
+	..()
```
Comment (MrDoomBringer) [inline code/game/gamemodes/objective.dm]:
```suggestion
	target = ..()
	if (target)
		to_chat(O.target.current, "<span class='hear'>You hear something crackle in your ear for a moment before a voice speaks. \"Please stand by for a message from the Syndicate. Message as follows: <b>Hey kid, we hear you got yourself in a bit of trouble with those Nanotrasen bigwigs. We're sending an agent down to try and bust you out of that place. Just remember to let our guy do the traitorous acts (like killing) - you don't need any more blood on your hands.</b> Message ends.\"</span>")

```

Comment (MrDoomBringer) [reply code/game/gamemodes/objective.dm]: You'd have to test it, but a solution like this should work.

**Comment by PKPenguin321** [comment]
so what if sec decides to just start killing prisoners once traitors are confirmed

File: code/game/gamemodes/objective.dm (hunk 2 above)
Comment (itseasytosee) [reply code/game/gamemodes/objective.dm]: oh I'm a dumbass

**Comment by MrDoomBringer** [comment]
@PKPenguin321 Other than the administration issue, that seems about as likely as like sec throwing away their reflector vest when traitors are confirmed, on the very off chance that it's an objective item.

**Comment by Shadowflame909** [comment]
Finally some use for the mulligan syringe.

Mulliganing your target doesn't break anything right?

**Comment by itseasytosee** [comment]
Nope. monkeyfying doesn't count ether not sure if that's an issue but I could fix it if it was.

**Comment by vpellen** [comment]
Roundstart prisoners should have a greentext for escaping out of custody

**Comment by RaveRadbury** [comment]
I think before we add anything to the prisoner's possibility for greentext we need to give Sec a greentext for bringing the prisoners onto the shuttle in custody, **_alive_**. Once that's set as a goal for sec then we can focus on antags and prisoners themselves having possible greentext.

The prisoner role exists to give sec something to do and encourage people to not go SSD/catatonic when perma'd. Anything that creates an incentive for killing prisoners (including suicides) should be preferably be avoided.

I support this PR as an admeme only objective. I'd like to see how security greentext plays out before it becomes a normal thing.

**Comment by itseasytosee** [comment]
Currently, it is 100% functional as admemes only which could lead to some fun stuff. I'm really not too comfortable adding this to traitor circulation and would be hoping that someone with a bit more trust in the community could call shots on that.

**Comment by Qustinnus** [comment]
can we make the prisoners play simon says

**Comment by itseasytosee** [comment]
what

**Comment by gitbirb** [comment]
> I think before we add anything to the prisoner's possibility for greentext we need to give Sec a greentext for bringing the prisoners onto the shuttle in custody, **_alive_**. Once that's set as a goal for sec then we can focus on antags and prisoners themselves having possible greentext.
> 
> The prisoner role exists to give sec something to do and encourage people to not go SSD/catatonic when perma'd. Anything that creates an incentive for killing prisoners (including suicides) should preferably be avoided.
> 
> I support this PR as an admeme only objective. I'd like to see how security greentext plays out before it becomes a normal thing.

I agree with this, currently the PR's suggested breakout objective probably wouldn't work since there is no incentive to keep prisoners whatsoever,  so prisoners always get paroled anyways and nothing happens of it because sec just doesn't want to bother.

**Comment by itseasytosee** [comment]
Ill add an insure X is in custody for admins to give to secmen.

**Comment by MrDoomBringer** [comment]
Why are you making this admin only?? This feature seems fine, and I doubt that many Admins are gonna remember to try this when it gets merged. Just add it as a normal roundstart objective

**Comment by itseasytosee** [comment]
It's just a general objective that would be used in tandem with the jailbreak objective.

**Comment by MrDoomBringer** [comment]
I get that, I'm saying why not make the jailbreak objective round start?

**Comment by itseasytosee** [comment]
Other people are worried about having sec instantly murder prisoners and stuff like that but I could probably make it happen.

**Comment by itseasytosee** [comment]
see rave RaveRadbury's points

**Comment by MrDoomBringer** [comment]
alright, yeah on second thought lets give prisoners some value to the station first

**Comment by RaveRadbury** [comment]
> Ill add an insure X is in custody for admins to give to secmen.

Is this being added roundstart or does an admin need to activate it.

I'm totally fine with the secmen greentext being roundstart. See how that plays out for a few days, pepper in the jailbreak a few times. After a week or two if it's worked out we can just phase jailbreak in as a standard antag greentext.

**Comment by itseasytosee** [comment]
It would need to be activated by admins

File: code/game/gamemodes/objective.dm [hunk 3]
```diff
@@ -306,6 +306,52 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
...
```
Comment (ShizCalev) [inline code/game/gamemodes/objective.dm]: return ..() ?

Comment (itseasytosee) [reply code/game/gamemodes/objective.dm]: Probably should.

File: code/game/gamemodes/objective.dm [hunk 4]
```diff
@@ -306,6 +306,52 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
 	name = "protect nonhuman"
 	human_check = FALSE
 
+/datum/objective/jailbreak
+	name = "jailbreak"
+	martyr_compatible = TRUE //why not?
+	var/target_role_type=FALSE
```
Comment (spookydonut) [inline code/game/gamemodes/objective.dm]:
```suggestion
	var/target_role_type
```

File: code/game/gamemodes/objective.dm [hunk 5]
```diff
@@ -306,6 +306,52 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
...
+	else
+		explanation_text = "Free Objective"
+
+/datum/objective/jailbreak/admin_edit(mob/admin)
+	admin_simple_target_pick(admin)
+
+/datum/objective/detain
+	name = "detain"
+	martyr_compatible = TRUE
+	var/target_role_type=FALSE
```
Comment (spookydonut) [inline code/game/gamemodes/objective.dm]:
```suggestion
	var/target_role_type
```

File: code/game/gamemodes/objective.dm [hunk 6]
```diff
@@ -306,6 +306,52 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
...
+	if(!invert)
+		target_role_type = role_type
+	return ..()
+
+/datum/objective/jailbreak/check_completion()
+	return completed || (considered_escaped(target))
+
+/datum/objective/jailbreak/update_explanation_text()
+	..()
+	if(target && target.current)
```
Comment (spookydonut) [inline code/game/gamemodes/objective.dm]:
```suggestion
	if(target?.current)
```

File: code/game/gamemodes/objective.dm [hunk 7]
```diff
@@ -306,6 +306,52 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
...
+	if(!invert)
+		target_role_type = role_type
+	return ..()
+
+/datum/objective/detain/check_completion()
+	return completed || (!considered_escaped(target) && (considered_alive(target) && target.current.onCentCom()))
+
+/datum/objective/detain/update_explanation_text()
+	..()
+	if(target && target.current)
```
Comment (spookydonut) [inline code/game/gamemodes/objective.dm]:
```suggestion
	if(target?.current)
```

**Review by spookydonut — CHANGES_REQUESTED** [review]
why cant you just have `/datum/objective/jailbreak` then `/datum/objective/jailbreak/escape` and `/datum/objective/jailbreak/detain`

The two objectives are basically identical code.

**Comment by itseasytosee** [comment]
Brilliant idea

**Comment by itseasytosee** [comment]
I'll do you one better and make the detain a child of jailbreak.

File: code/game/gamemodes/objective.dm [hunk 8]
```diff
@@ -306,6 +306,42 @@ GLOBAL_LIST(admin_objective_list) //Prefilled admin assignable objective list
 	name = "protect nonhuman"
 	human_check = FALSE
 
+/datum/objective/jailbreak
+	name = "jailbreak"
+	martyr_compatible = TRUE //why not?
+	var/target_role_type
+
+/datum/objective/jailbreak/find_target_by_role(role, role_type=FALSE,invert=FALSE)
```
Comment (spookydonut) [inline code/game/gamemodes/objective.dm]:
```suggestion
/datum/objective/jailbreak/find_target_by_role(role, role_type, invert=FALSE)
```

**Review by spookydonut — CHANGES_REQUESTED** [review]
looks good!  have you tested this?

**Comment by itseasytosee** [comment]
I have tested this thoroughly with getting an alt account on my laptop. Both objectives fail if the target is marooned the detain on succeeds if they are in the sec section and it's not hijacked jailbreak works as long as they are alive, "escaped" on the Centcom z-level or whiteship and not in the sec zone.

**Review by spookydonut — APPROVED** [review]

**Pull request merged**